components/data-feed-service/service/data-feed-workflow.go

```diff
@@ -0,0 +1,157 @@
+package service
+
+import (
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/chef/automate/lib/cereal"
+)
+
+var (
+	dataFeedWorkflowName = cereal.NewWorkflowName("data-feed-workflow")
+	dataFeedScheduleName = "periodic-data-feed-workflow"
+)
+
+type DataFeedWorkflowExecutor struct {
+	workflowName cereal.WorkflowName
+}
+
+// DataFeedWorkflowParams the params for the workflow and all the tasks in the workflow
+type DataFeedWorkflowParams struct {
+	UpdatedNodesOnly bool
+	NodeBatchSize    int
+	NodeIDs          map[string]NodeIDs
+	FeedStart        time.Time
+	FeedEnd          time.Time
+	PollTaskParams   DataFeedPollTaskParams
+}
+
+type DataFeedWorkflowPayload struct {
+	TasksComplete bool
+	NodeIDs       []map[string]NodeIDs
+	NodeIDsIndex  int
+}
+
+func (e *DataFeedWorkflowExecutor) OnStart(w cereal.WorkflowInstance, ev cereal.StartEvent) cereal.Decision {
+	params := DataFeedWorkflowParams{}
+	log.Debugf("OnStart params %v", params)
+	err := w.GetParameters(&params)
+	if err != nil {
+		return w.Fail(err)
+	}
+
+	/*
+	 * Start the workflow with the poll task to get any nodes that have had client runs
+	 */
+	err = w.EnqueueTask(dataFeedPollTaskName, params)
+	if err != nil {
+		return w.Fail(err)
+	}
+
+	initialPayload := DataFeedWorkflowPayload{}
+	return w.Continue(&initialPayload)
+}
+
+func (e *DataFeedWorkflowExecutor) OnTaskComplete(w cereal.WorkflowInstance, ev cereal.TaskCompleteEvent) cereal.Decision {
+	params := DataFeedWorkflowParams{}
+	err := w.GetParameters(&params)
+	if err != nil {
+		return w.Fail(err)
+	}
+
+	log.Debugf("OnTaskComplete params: %v", params)
+	payload := DataFeedWorkflowPayload{}
+	err = w.GetPayload(&payload)
+	if err != nil {
+		return w.Fail(err)
+	}
+
+	switch ev.TaskName {
+	case dataFeedPollTaskName:
+		taskResults, err := getPollTaskResults(ev)
+		if err != nil {
+			return w.Fail(err)
+		}
+		/*
+		 * Update the workflow params
+		 */
+		params.NodeIDs = taskResults.NodeIDs
+		if len(taskResults.NodeIDs) > 0 {
+			payload.NodeIDs = batchNodeIDs(params.NodeBatchSize, taskResults.NodeIDs)
+			log.Debugf("payload.NodeIDs %v payload.NodeIDsIndex %v", payload.NodeIDs, payload.NodeIDsIndex)
+			params.NodeIDs = payload.NodeIDs[payload.NodeIDsIndex]
+			err = w.EnqueueTask(dataFeedAggregateTaskName, params)
```

Comment:
not that this is wrong or that you should change it, but you could enqueue all the batches and have only one thread available to process them per machine. Unless you need exactly one batch to be processed at a time, then for sure leave it as is

Reply:
Simply doing them in serial in a first pass to see how we perform. We don't need exactly one batch to be processed at a time unless we only have 1 machine.